pkg/doc/jwt/claims.go

```diff
@@ -0,0 +1,123 @@
+/*
+ *
+ *  * Copyright SecureKey Technologies Inc. All Rights Reserved.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package jwt
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+	"time"
+)
+
+// Claims defines JWT Claims Set (https://tools.ietf.org/html/rfc7519#section-4)
+type Claims struct {
+
+	// Issuer defines iss claim.
+	Issuer string `json:"iss,omitempty"`
+
+	// Subject defines sub claim.
+	Subject string `json:"sub,omitempty"`
+
+	// Audience defines aud claim.
+	Audience Audience `json:"aud,omitempty"`
+
+	// Expiry defines exp claim.
+	Expiry *NumericDate `json:"exp,omitempty"`
+
+	// NotBefore defines nbf claim.
+	NotBefore *NumericDate `json:"nbf,omitempty"`
+
+	// IssuedAt defines iat claim.
+	IssuedAt *NumericDate `json:"iat,omitempty"`
+
+	// ID defines jti claim.
+	ID string `json:"jti,omitempty"`
+}
+
+// NumericDate is a JSON numeric value representing the number of seconds from
+// 1970-01-01T00:00:00Z UTC until the specified UTC date/time, ignoring leap seconds.
+type NumericDate int64
+
+// NewNumericDate creates a new NumericDate from time.Time
+func NewNumericDate(t time.Time) *NumericDate {
+	if t.IsZero() {
+		return nil
+	}
+
+	nd := NumericDate(t.Unix())
+
+	return &nd
+}
+
+// MarshalJSON converts NumericDate to JSON bytes.
+func (n NumericDate) MarshalJSON() ([]byte, error) {
+	return []byte(strconv.FormatInt(int64(n), 10)), nil
+}
+
+// UnmarshalJSON defines custom unmarshalling of NumericDate from JSON bytes.
+func (n *NumericDate) UnmarshalJSON(data []byte) error {
+	s := string(data)
+
+	f, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		return errors.New("not a number value")
+	}
+
+	*n = NumericDate(f)
+
+	return nil
+}
+
+// Time provides time.Time value of NumericDate.
+func (n NumericDate) Time() time.Time {
+	return time.Unix(int64(n), 0)
+}
+
+// Audience identifies the recipients that the JWT is intended for.
+type Audience []string
```

Comment:
The below code doesn't look "Audience" specific. Why not have a generic type for a String slice that can unmarshal both a single string element or an array of strings?

Reply:
Makes sense.